File: year_2025/src/day_3/day_3_2.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int find_max(char *s, int l, int r) {
  int max_val = -1;
  int max_pos = 0;

  for (int i = l; i < r; i++) {
    if (s[i] > max_val) {
      max_val = s[i];
      max_pos = i;
    }
  }

  return max_pos;
}

int main() {
  FILE *file = fopen("data/day_3/input", "r");

  long res = 0;

  char line[1024];
  while (fgets(line, sizeof(line), file)) {
    line[strcspn(line, "\n")] = '\0';
    printf("%s\n", line);

    int idx = -1;
    long jolts = 0;
    for (int i = 11; i >= 0; i--) {
      idx = find_max(line, idx + 1, strlen(line) - i);
      long num = line[idx] - '0';
      printf("%d\t%ld\n", idx, num);
      jolts += num * (long)pow(10, i);
    }
    printf("%ld\n", jolts);

    res += jolts;
  }

  printf("%ld\n", res);

  fclose(file);
}

File: year_2025/src/day_3/day_3_2.py
def main():
    with open("data/day_3/input", "r") as f:
        lines = f.readlines()
        lines = [line.split("\n")[0] for line in lines]

    res = 0
    for line in lines:
        stack = []
        for i, c in enumerate(line):
            digit = int(c)

            num_remaining = len(line) - i
            while len(stack) + num_remaining > 12 and stack and digit > stack[-1]:
                stack.pop()
            stack.append(digit)

        jolt = sum(d * pow(10, 12 - i) for i, d in enumerate(stack[:12], start=1))
        res += jolt

    print(res)


if __name__ == "__main__":
    main()
